### ISSUE TITLE
scrollBehaviour should be executed before after hooks when pressing the back button

### ISSUE THREAD

**Issue by narusas** [vuejs/vue-router#2095]
### Version
3.0.1

### Reproduction link
[https://jsfiddle.net/narusas/ektba5vj/8/](https://jsfiddle.net/narusas/ektba5vj/8/)

### Steps to reproduce
Open https://jsfiddle.net/narusas/ektba5vj/8/ 
Open Dev console
Click '/hello1'
Clear console

Scroll some
Check console 
```javascript
	scroll 4
 	scroll 8
 	scroll 12
```

Click '/hello2'
Check console
```javascript
	Component1 beforeRouteLeave
	next-beforeRouteLeave-Component1
 	global router  beforeEach
 	/hello2 route beforeEnter
 	Component2 beforeRouteEnter
	next-beforeRouteEnter-Component2
	global router beforeResolve
	global router afterEach
	Component2 beforeCreate
	Component2 created
	Component2 beforeMount
	Component2 mounted
	scrollBehavior
	scrollBehavior - resolve
	scroll 0 
 ```

Currently It's ok

Now press back-button
Check console
```javascript
	Component2 beforeRouteLeave
	scroll 12   <---- Why This Timing???? Why leaving component is  scrolled??  
	next-beforeRouteLeave-Component2
	global router  beforeEach
	/hello1 route beforeEnter
	Component1 beforeRouteEnter
	next-beforeRouteEnter-Component1
	global router beforeResolve
	global router afterEach
	Component1 beforeCreate
	Component1 created
	Component1 beforeMount
	Component1 mounted
	scrollBehavior
```

### What is expected?
When press back button, 
Before router leave, no scrolling is executed.

### What is actually happening?
When press back button,
Leave component is  scrolled strange position. 
After that scrolling, Entering Component is entered.  
It looks like flickering. 

---
Now I creating some SPA web site. 
When user press back/foward button, leaving component is scrolled and after that show entering component. 
SO It is flickering!
Also, I use lazy image loading. So user press back/foward, this not-intented  scrolling will call unnecessary image loading , and progressive loading animation is showed.

**Comment by posva** [comment]
Changed the title to _scrollBehaviour should be executed before after hooks when pressing the back button_

**Comment by posva** [comment]
This got fixed with 12613638a4cea74b5f205c6dfa314e96fbd9817a
